### ISSUE TITLE
U-0081 is not in unicode_db.c. It's NULL

### ISSUE THREAD

**Issue by nwc10** [MoarVM/MoarVM#1322]
In `unicode_db.c` we have this:

```
    /*125*/"RIGHT CURLY BRACKET"/* 007D */,/*126*/"TILDE"/* 007E */,/*127*/"<control>"/* 007F */,
    /*128*/"<control>"/* 0080 */,NULL,/*130*/"<control>"/* 0082 */,/*131*/"<control>"/* 0083 */,/*132*/"<control>"/* 0084 */,
    /*133*/"<control>"/* 0085 */,/*134*/"<control>"/* 0086 */,/*135*/"<control>"/* 0087 */,/*136*/"<control>"/* 0088 */,
```

I can't work out *how* 129 ends up as `NULL` when everything else is correct. There doesn't seem to be any special-casing in `ucd2c.pl` that would cause it, and as best I can tell, 0x0081 features in all the same places as 0x0080 and 0x0082. For example,  `NamesList.txt`

```
@@      0080    C1 Controls and Latin-1 Supplement (Latin-1 Supplement) 00FF
@               C1 controls
@+              Alias names are those for ISO/IEC 6429:1992.
0080    <control>
0081    <control>
0082    <control>
        = BREAK PERMITTED HERE
        x (zero width space - 200B)
```

**Comment by AlexDaniel** [comment]
So what's the issue here? Can you demonstrate it with a code snippet?

**Comment by nwc10** [comment]
The issue is that that section of the generated file `unicode_db.c` is not correct. It should read this;

```
    /*125*/"RIGHT CURLY BRACKET"/* 007D */,/*126*/"TILDE"/* 007E */,/*127*/"<control>"/* 007F */,
    /*128*/"<control>"/* 0080 */,/*129*/"<control>"/* 0081 */,/*130*/"<control>"/* 0082 */,/*131*/"<control>"/* 0083 */,/*132*/"<control>"/* 0084 */,
    /*133*/"<control>"/* 0085 */,/*134*/"<control>"/* 0086 */,/*135*/"<control>"/* 0087 */,/*136*/"<control>"/* 0088 */,
```

because for that array, index 129 should have that value (to be consistent)

The file is generated by `ucd2c.pl` from the Unicode character database - there is nothing in those files which treats 0080 differently from 0081, so there has to be some sort of strange bug in our code that reads those files and generates these C file from them.

As to a code snippet, I suspect that the data in this table is actually completely unreachable (see #1321) because we are (also) currently trying to assign the (repeated) values from this table, which doesn't "work" - we are assigning duplicate keys, but with different values.

**Comment by nwc10** [comment]
It was caused by logic in `ucd2c.pl` and the C code intended to be an optimisation. It didn't actually affect the return values for calls related to U-0081, but in brute force testing:

```
./nqp-m -e 'my $i := 0; while $i < 0x120000 { my $n := nqp::getuniname($i); my $raw := nqp::codepointfromname($n); my $cp := $raw == -1 ?? "-" !! nqp::sprintf("%X", [$raw]); nqp::say(nqp::sprintf("%5X %6s %s    %s", [$i, $cp, (($raw == -1 || $raw == $i) ?? " " !! "!"), $n])); ++$i; }'
```

I discovered that there were 64 code points that are actually (still) `<reserved>` because they are not yet allocated by Unicode, but we had been reporting them as allocated in the block that they were in. (see commit 07dbc19a5e4ba56fe474e9136b8e407990486290)

Fixed as part of the fixes for #1321
